### ISSUE TITLE
pytorch study 8주차

### ISSUE THREAD

**Issue by hyunsu-yang** [hyunsu-yang/PyTorch#10]
## 강의
- https://www.edwith.org/boostcourse-dl-pytorch/joinLectures/24017

## 목차 
- Lab-10-4-1 ImageFolder1
- Lab-10-4-2 ImageFolder2
- Lab-10-5 Advance CNN(VGG)

**Comment by hyunsu-yang** [comment]
## Lab-10-4-1 ImageFolder1

#### 핵심
- image Folder
```python
// transform.Compose() 이미지 변환 수행할 때 하나로 묶어주는 역할
trans = transforms.Compose([
    transforms.Resize((64,128))
])

train_data = torchvision.datasets.ImageFolder(root='custom_data/origin_data', transform=trans)
```

**Comment by hyunsu-yang** [comment]
## Lab-10-4-1 ImageFolder2

#### 핵심
```python
# 이미지 학습된 모델을 중간 저장하는 방법
torch.save(net.state_dict(), "./model/model.pth")

# 새로운 네트워크에 기학습된 모델 불러오기
new_net.load_state_dict(torch.load('./model/model.pth'))

# 두모델 같은지 비교 방법
print(net.layer1[0])
print(new_net.layer1[0])

print(net.layer1[0].weight[0][0][0])
print(new_net.layer1[0].weight[0][0][0])

net.layer1[0].weight[0] == new_net.layer1[0].weight[0]
```

**Comment by hyunsu-yang** [comment]
## Lab-10-5 Advance CNN(VGG)

#### 핵심
- kaiming_normal_ ?
  - kaming normalize 해줘라, 다양한 activation fuction을 이용해서 normalize  해줘라..

- 나만의 VGG_NET 만드는 방법
```python
# 'A': [64, 'M', 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M']
    
def make_layers(cfg, batch_norm=False):
    layers = []
    in_channels = 3
    
    for v in cfg:
        if v == 'M':
            layers += [nn.MaxPool2d(kernel_size=2, stride=2)]
        else:
            conv2d = nn.Conv2d(in_channels, v, kernel_size=3, padding=1)
            if batch_norm:
                layers += [conv2d, nn.BatchNorm2d(v), nn.ReLU(inplace=True)]
            else:
                layers += [conv2d, nn.ReLU(inplace=True)]
            in_channels = v
                     
    return nn.Sequential(*layers)

cfg = {
    'A': [64, 'M', 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'], #8 + 3 =11 == vgg11
    'B': [64, 64, 'M', 128, 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'], # 10 + 3 = vgg 13
    'D': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 'M', 512, 512, 512, 'M', 512, 512, 512, 'M'], #13 + 3 = vgg 16
    'E': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 256, 'M', 512, 512, 512, 512, 'M', 512, 512, 512, 512, 'M'], # 16 +3 =vgg 19
    'custom' : [64,64,64,'M',128,128,128,'M',256,256,256,'M']
}

conv = make_layers(cfg['custom'], batch_norm=True)
CNN = VGG(make_layers(cfg['custom']), num_classes=10, init_weights=True)
```
- out
```
VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): ReLU(inplace)
    (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (3): ReLU(inplace)
    (4): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (5): ReLU(inplace)
    (6): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (7): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): ReLU(inplace)
    (9): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (10): ReLU(inplace)
    (11): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (12): ReLU(inplace)
    (13): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (14): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (15): ReLU(inplace)
    (16): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (17): ReLU(inplace)
    (18): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (19): ReLU(inplace)
    (20): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (avgpool): AdaptiveAvgPool2d(output_size=(7, 7))
  (classifier): Sequential(
    (0): Linear(in_features=25088, out_features=4096, bias=True)
    (1): ReLU(inplace)
    (2): Dropout(p=0.5)
    (3): Linear(in_features=4096, out_features=4096, bias=True)
    (4): ReLU(inplace)
    (5): Dropout(p=0.5)
    (6): Linear(in_features=4096, out_features=10, bias=True)
```

**Comment by hyunsu-yang** [comment]
# 문제
## Lab-10-4 ImageFolder
- pytorch imagefolder에 대해 간단히 설명하세요. 
  - 계층적인 폴더 구조를 가지고 있는 데이터셋을 불러올 때 사용
  - 각 이미지들이 자신의 레이블(Label) 이름으로 된 폴더 안에 들어가 있는 구조라면,
     ImageFolder 라이브러리를 이용하여 이를 바로 불러와 객체로 만들면 됨.
```

dataset/ 
  0/ 
    0.jpg 
    1.jpg ...  
  1/ 
    0.jpg 
    1.jpg ... 
  ... 
  9/ 
    0.jpg 
    1.jpg
```



## Lab-10-5 Advance CNN(VGG)
- padding=0이고, stride=1 일때, 3 * 3 conv filter를 두번 사용하는 것과 5 * 5 conv filter를 한번 사용하는 것은 같은 결과를 갖는다. 그런데 VGGNet에서 3*3 conv filter를 두번 사용하는 이유는? 
  - 더 많은 conv filter를 사용하게 되면 여러개의 non linear function인 ReLU를 더 많이 사용할 수 있게되어
     non-linear한 문제를 더 잘 다룰수 있는 장점
![image](https://user-images.githubusercontent.com/19383191/98615443-ff694b80-233d-11eb-847a-47a6541072c1.png)
  - 연산시 적용되는 parameter수가 줄어든다는 점,  
    이렇게 줄어든 parameter들은 overfitting을 줄여줌.
